rpki-validator/src/main/java/net/ripe/rpki/validator3/rrdp/RrdpServiceImpl.java

```diff
@@ -116,23 +124,16 @@ private boolean doStoreRepository(RpkiRepository rpkiRepository, RpkiRepositoryV
         if (notification.sessionId.equals(rpkiRepository.getRrdpSessionId())) {
             if (rpkiRepository.getRrdpSerial().compareTo(notification.serial) <= 0) {
                 try {
-                    final List<Delta> deltas = notification.deltas.stream().
-                            filter(d -> d.getSerial().compareTo(rpkiRepository.getRrdpSerial()) > 0).
-                            sorted(Comparator.comparing(DeltaInfo::getSerial)).
-                            parallel().
-                            map(di -> readDelta(notification, di)).
-                            collect(Collectors.toList());
-
-                    verifyDeltaSerials(deltas, notification, rpkiRepository);
-
-                    deltas.forEach(d -> {
-                        storage.readTx0(tx -> verifyDeltaIsApplicable(tx, d));
-                        storage.writeTx0(tx -> {
-                            storeDelta(tx, d, validationRun, rpkiRepository, changedObjects);
-                            tx.afterCommit(() -> rpkiRepository.setRrdpSerial(rpkiRepository.getRrdpSerial().add(BigInteger.ONE)));
-                        });
-                    });
-
+                    List<DeltaInfo> orderedDeltas = verifyDeltaSerials(notification, rpkiRepository);
+                    rrdpProcessingPool.submit(() -> orderedDeltas
+                            .parallelStream()
+                            .map(di -> readDelta(notification, di))
+                            .forEachOrdered(d -> storage.writeTx0(tx -> {
```

Comment:
A bit confused with combination of forEachOrdered with parallelStream. Wouldn't it cancel the parallel effect?

Reply:
Downloads are now happening concurrently, which is usually IO bound. But the final committing to the data is done one-by-one, in order, after the first download completes, then the next, etc. Out-of-order completion of downloads is handled correctly this way.